nntrainer/src/databuffer_v2.cpp

```diff
@@ -0,0 +1,231 @@
+//     batched_label_holder = allocateBatchedDataHolder(label_size);
+//
+//     batched_buffer.push_back(std::make_pair(batched_input_holder,
+//     batched_label_holder)); for (unsigned int b = 0; b < batch_size; b++) {
+//       unsigned int buf_idx = batched_buflen * batch_size + b;
+//       std::vector<void *> input_holder;
+//       for (auto const in&: batched_input_holder):
+//         input_holder.push_back(invVoidPtr(in[]))
+//       buffer.push_back(std::make_pair(input_holder));
+//     }
+// }
+
+void DataBuffer_v2::setDataSource(datagen_cb gen_cb, void *user_data) {
+  if (type == DataBufferType::DATA_BUFFER_FILE)
+    throw std::invalid_argument("setting generator for file based dataset");
+
+  if (gen_cb == nullptr)
+    throw std::invalid_argument("callback function is null");
+
+  if (generator)
+    ml_logw("Overwriting existing data generator source.");
+
+  type = DataBufferType::DATA_BUFFER_GENERATOR;
+  generator = gen_cb;
+  gen_user_data = user_data;
+}
+
+void DataBuffer_v2::setDataSource(const std::string file) {
+  DataBufferFileUserData *file_usr_data;
+  if (type == DataBufferType::DATA_BUFFER_GENERATOR)
+    throw std::invalid_argument("setting file for generator based dataset");
+
+  if (gen_user_data) {
+    ml_logw("Overwriting existing data file source.");
+    file_usr_data = (DataBufferFileUserData *)gen_user_data;
+    delete file_usr_data;
+  }
+
+  type = DataBufferType::DATA_BUFFER_FILE;
+  generator = file_dat_cb;
+  file_usr_data = new DataBufferFileUserData(file);
+  gen_user_data = static_cast<void *>(file_usr_data);
+}
+
+void DataBuffer_v2::init() {
+  if (type == DataBufferType::DATA_BUFFER_UNKNOWN)
+    throw std::runtime_error("Data source has not been set for the dataset");
+
+  for (auto const &size : input_size)
+    if (size == 0)
+      throw std::invalid_argument("Input size is not set for all the elements");
+
+  for (auto const &size : label_size)
+    if (size == 0)
+      throw std::invalid_argument("Label size is not set for all the elements");
+
+  if (type == DataBufferType::DATA_BUFFER_FILE) {
+    DataBufferFileUserData *file_usr_data =
+      (DataBufferFileUserData *)gen_user_data;
+    file_usr_data->setInputLabelSize(input_size, label_size);
+    total_data_entries =
+      file_usr_data->size / file_usr_data->getSingleDataSize();
+  }
+
+  if (type == DataBufferType::DATA_BUFFER_FILE &&
+      batch_size > total_data_entries)
+    throw std::runtime_error("Batch size exceeds the total datasize");
+
+  if (buffer_len > total_data_entries)
+    buffer_len = total_data_entries;
+
+  if (buffer_len < batch_size)
+    buffer_len = batch_size;
+
+  /**
+   * batch size data is allocated at once in the buffer for data to avoid
+   * copying to data back into another holder which has contiguous data.
+   * This limits the buffer length to be a multiple of batch size.
+   */
+  batched_buffer_len = (buffer_len / batch_size);
+  buffer_len = batched_buffer_len * batch_size;
```

Comment:
I think this is duplication. buffer_len is used to set batched_buffer_len and set the buffer_len using this.